### PR TITLE
feat: add support for certificates

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
**Description**
This PR allows eox-tagging to tag GeneratedCertificate objects through the eox-tagging API. Using the tag model they can also be tagged passing the object itself.

**How to test**
First, add the following configuration:

```
EOX_TAGGING_DEFINITIONS = [
    {
        "owner_object": "Site",
        "tag_type": "cert_dates",
        "validate_target_object": "GeneratedCertificate"
    }
]
```

**Create a tag using the Tagging API:**

```
{
    "tag_type": "cert_dates",
    "tag_value": "2020-09-09",
    "target_type": "generatedcertificate",
    "target_id": "edx: course-v1:edX+E2E-101+course",
    "owner_type": "site"
}
```
or
```
{
  "tag_value": "2020",
  "tag_type": "cert_dates",
  "target_id": "0441c66eb13d47deb3c79c79027f822b",
  "target_type": "generatedcertificate"
}
```
You can use as `target_id` the format `username: course_id` or the verify_uuid of the certificate

**Get tag for specific course_id and username:**
image

image

**Ger certificate by verify_uuid**
image

### CODE REVIEW

File: eox_tagging/api/v1/filters.py [hunk 1]
```diff
@@ -26,20 +29,42 @@ class Meta:
...
+            value = int(value) if name == "id" else value
+        return name, value
+
     def filter_by_target_object(self, queryset, name, value):
         """Filter that returns the tags associated with target."""
-        TARGET_TYPES = {
+        target_types = {
             "course_id": "courseoverview",
             "username": "user",
+            "certificate_id": "generatedcertificate",
```
Comment (mariajgrimaldi) [inline eox_tagging/api/v1/filters.py:55]: I just realized we did something similar with courseenrollment. I'll be changing this implementation to match that one

File: eox_tagging/api/v1/filters.py [hunk 2]
```diff
@@ -26,20 +29,42 @@ class Meta:
...
+            "certificate_key": "generatedcertificate",
+            "certificate_url": "generatedcertificate",
         }
         if value:
-            DEFAULT = {
-                name: value,
+            model_field_name, model_value = self._map_to_model(name, value)
```
Comment (felipemontoya) [inline eox_tagging/api/v1/filters.py:60]: This call to the _map_to_model function looks like it is trying to abstract some logic that should just be explicit.

**Review by felipemontoya — CHANGES_REQUESTED** [review]

File: eox_tagging/api/v1/filters.py (hunk 2 above)
Comment (mariajgrimaldi) [reply eox_tagging/api/v1/filters.py:60]: I can change it to `parse_search_string`

Comment (felipemontoya) [reply eox_tagging/api/v1/filters.py:60]: We have an explicit mapping donde vía the `target_types` dict and a implicit mapping done via the `_map_to_model` function that looks at the string and returns `int(value)` for "certificate_id" and or `value` otherwise.

Thats is really convoluted. We should go for one of the two. Either a function that returns the `filter_params` for different all types vía a long mapping dictionary, or something that does different functions for each type of target_type and we call each function by looking at the target_type and calling one for course_id, other for username and other for certificates, ...

Comment (felipemontoya) [reply eox_tagging/api/v1/filters.py]: I had already published the other comment when I noticed yours @mariajgrimaldi.

I don't mean the function name. I mean the actual function. We either do it implicitly (a function that does `if something then`) or explicitly (a big dict with all the mapping in it) but doing one for some types of targets and something different for certificates is plain bad practice.

File: eox_tagging/api/v1/filters.py [hunk 3]
```diff
@@ -26,22 +36,36 @@ class Meta:
...
         fields = ["key", "status", "tag_type", "tag_value"]
 
+    @staticmethod
+    def _map_to_filter_params(search_field, value):
+        """
+        Used to parse search strings to fields/values related to the queryset to search.
+
+        Example usage:
+            >>> filter_params = self._parse_search_string("certificate_id", "1")
+            >>> {"target_id": {"id": 1}, "target_type": ""generatedcertificate"}
```
Comment (felipemontoya) [inline eox_tagging/api/v1/filters.py:46]: typo double ""

File: eox_tagging/api/v1/filters.py [hunk 4]
```diff
@@ -26,22 +36,36 @@ class Meta:
...
+        """
+        filter_params = {
+            "target_type": FILTER_TARGET_TYPES.get(search_field),
+        }
+        if search_field.startswith("certificate_"):
```
Comment (felipemontoya) [inline eox_tagging/api/v1/filters.py:51]: this whole block still is super convoluted.
Basically you want `value` for every single case but `certificate_id`, why dont you make the if be `if name ='certificate_id` and be done with it?

**Review by felipemontoya — DISMISSED** [review]

File: eox_tagging/api/v1/filters.py (hunk 3 above)
Comment (mariajgrimaldi) [inline eox_tagging/api/v1/filters.py:46]:
```suggestion
            >>> {"target_id": {"id": 1}, "target_type": "generatedcertificate"}
```

File: eox_tagging/api/v1/filters.py (hunk 4 above)
Comment (mariajgrimaldi) [reply eox_tagging/api/v1/filters.py:51]: Because we need to search the queryset with `id` and not `certificate_id`

Comment (felipemontoya) [reply eox_tagging/api/v1/filters.py:51]: Yes, I see that, but you are still making it convoluted for no reason.

For 4 fields you need:

```
"course_id": -> {"course_id": value}
"username": -> {"username": value}
"certificate_key": -> {"certificate_key": value}
"certificate_url": -> {"certificate_url": value}
```

but the last field requires
```
"certificate_id": -> {"id": int(value)}
```

This is why I ask: why not make it:

```
if search_field in ["certificate_id"]:
            filter_params["target_id"] = {
                "id": int(value)
            }
else:
            filter_params["target_id"] = {
                search_field: value,
            }

```

Which makes it clear about which fields require something special.
Right now,to understand the code I need to imagine the logic of this block on all the possible fields just so 1 of them can be singled out.

Comment (mariajgrimaldi) [reply eox_tagging/api/v1/filters.py:51]: Thanks for the explanation! Now I understand better what you're saying.

The thing is we need to do the _transformation_ with all the fields that start with `certificate`: certificate_download_url -> download_url, certificate_key -> key, certificate_id -> id. So the `target_id` can be:

1. `{"id": int(value)}`
2. `{"key": value}`
3. `{"download_url":value}`

I'll be checking again tomorrow to see if I miss something.

Comment (mariajgrimaldi) [reply eox_tagging/api/v1/filters.py:51]: The _transformation_ also includes:

"certificate_key": -> {"key": value}
"certificate_download_url": -> {"download_url": value}

I'll be checking this tomorrow to see if we can accomplish something better.

File: eox_tagging/edxapp_accessors.py [hunk 5]
```diff
@@ -55,16 +58,33 @@ def get_course_enrollment(**kwargs):
...
+    username = kwargs.get("username")
+    course_id = kwargs.get("course_id")
+
+    return GeneratedCertificate.objects.get(
+        user__username=username,
+        course_id=CourseKey.from_string(course_id),
+    )
+
+
 def get_object(related_object_type, **kwargs):
```
Comment (MoisesGSalas) [inline eox_tagging/edxapp_accessors.py:75]: This functions is really hard to parse.

You can maybe rename it to `get_related_object` an rename the arg to `type` then use a chained `if..elif` block to call the correct getter.
```python
if type == 'user':
    get_user(**kwargs)
elif type == 'courseoverview':
    get_course(**kwargs)
elif type == 'site':
    get_site(**kwargs)
elif type == 'courseenrollment':
    get_course_enrollment(**kwargs)
elif type == 'generatedcertificate':
    get_certificate(**kwargs)
```

File: eox_tagging/api/v1/serializers.py [hunk 6]
```diff
@@ -58,7 +59,7 @@ def create(self, validated_data):
         target = validated_data.pop("target_object", None)
 
         if target_type and target_type.lower() in MODELS_WITH_COMPOUND_KEYS:
-            data = self.__convert_compound_keys(target, target_type)
+            data = self.__convert_compound_keys(target, target_type.lower())
```
Comment (MoisesGSalas) [inline eox_tagging/api/v1/serializers.py:62]: About the `__convert_compound_keys`, the separator has to be exactly `: `  (i.e.1 colon **_and_** 1 space)? Can you make it a bit more forgiving? Ignoring the spaces.

File: eox_tagging/edxapp_accessors.py [hunk 7]
```diff
@@ -55,16 +58,33 @@ def get_course_enrollment(**kwargs):
...
         "site": get_site,
         "courseenrollment": get_course_enrollment,
+        "generatedcertificate": get_certificate,
     }
     try:
-        related_object = RELATED_OBJECTS.get(related_object_type.lower())(**kwargs)
+        related_object = related_objects_getter.get(
+            related_object_type.lower(),
+        )(**kwargs)
     except Exception:
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply eox_tagging/edxapp_accessors.py]: I removed catching the exception. If an exception is raised when creating a tag, then it is captured by the serializer. And if another service uses it, then it's better that raises the correct exception instead of ValidationError

File: eox_tagging/api/v1/serializers.py [hunk 8]
```diff
@@ -88,11 +82,16 @@ def create(self, validated_data):
...
             raise serializers.ValidationError({"Tag": _("{}".format(e.message))})
 
-    def __convert_compound_keys(self, ids, object_type):
+    @staticmethod
+    def _convert_compound_keys(ids, object_type):
         """
         Function that converts strings with format: `key1: key2` into a dictionary.
         """
-        target_id = re.split(r':\s', ids)
+        target_id = ids.replace(" ", "").split(":", 1)
```
Comment (MoisesGSalas) [inline eox_tagging/api/v1/serializers.py:90]: This applies to enrollments too right? shouldn't it be `rsplit` instead? or I'm missing something?

Comment (mariajgrimaldi) [reply eox_tagging/api/v1/serializers.py:90]: I don't get it. Course enrollments use: `username: course_id`

Comment (MoisesGSalas) [reply eox_tagging/api/v1/serializers.py:90]: My bad, I had it backwards in my mind `course_id: username`.

**Review by MoisesGSalas — APPROVED** [review]
